### PR TITLE
Parallelized subseq function

### PR DESCRIPTION
**Pull request by CodingKaiser**

### CODE REVIEW

File: src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java [hunk 1]
```diff
@@ -50,107 +50,107 @@ public SubsequenceFunctionIterator(
...
         super(parameters, executionMode, iteratorMetadata);
-    }
-
-    @Override
-    public void open(DynamicContext context) {
-        super.open(context);
+        this.sequenceIterator = this.children.get(0);
 
-        this.currentPosition = 1; // JSONiq indices start from 1
+        Item positionItem = this.children.get(1)
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java]: The constructor of an iterator cannot manipulate children iterators -- this must be done in open() (for local), and in getRDDAux() (for parallel). This implementation is likely to break as soon as you provide a variable for parameters 2 and 3.

File: src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java [hunk 2]
```diff
@@ -50,107 +50,107 @@ public SubsequenceFunctionIterator(
...
-                        "Invalid args. subsequence can't be performed with an object parameter as the length",
-                        getMetadata()
-                );
-            } else if (!(lengthItem.isNumeric())) {
-                throw new UnexpectedTypeException(
-                        "Invalid args. Length parameter should be an numeric(Integer/Decimal/Double)",
-                        getMetadata()
+    @Override
+    protected JavaRDD<Item> getRDDAux(DynamicContext context) {
+        JavaRDD<Item> childRDD = this.sequenceIterator.getRDD(context);
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java]: This implementation looks good.

File: src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java [hunk 3]
```diff
@@ -50,107 +50,107 @@ public SubsequenceFunctionIterator(
...
+    @Override
+    protected void closeLocal() {
+        this.sequenceIterator.close();
+    }
 
+    @Override
+    protected void resetLocal(DynamicContext context) {
+        int currentPosition = 1; // JSONiq indices start from 1
+
+        this.currentLength = this.length;
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java]: in the future, the code in reset and open, which is duplicate, could be factored out as an init method.

File: src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java (hunk 1 above)
Comment (CodingKaiser) [reply src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java]: I see. This should now be fixed.

File: src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java (hunk 3 above)
Comment (CodingKaiser) [reply src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java]: I have fixed this in the latest version

File: src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java [hunk 4]
```diff
@@ -50,107 +52,92 @@ public SubsequenceFunctionIterator(
             ExceptionMetadata iteratorMetadata
     ) {
         super(parameters, executionMode, iteratorMetadata);
+        this.sequenceIterator = this.children.get(0);
+        this.positionIterator = this.children.get(1);
+        if (this.children.size() == 3) {
+            this.lengthIterator = this.children.get(2);
+        }
```
Comment (ghislainfourny) [inline src/main/java/org/rumbledb/runtime/functions/sequences/general/SubsequenceFunctionIterator.java]: Perfect

**Comment by CodingKaiser** [comment]
I've added two additional tests for this. I'll make sure to do this going forward.

**Pull request merged**